lms/services/grouping.py

```diff
@@ -32,52 +33,49 @@ def generate_authority_provided_id(
             tool_consumer_instance_guid, parent.lms_id, type_.value, lms_id
         )
 
-    def upsert_with_parent(  # pylint: disable=too-many-arguments
-        self,
-        tool_consumer_instance_guid,
-        lms_id,
-        lms_name,
-        parent: Grouping,
-        type_: Grouping.Type,
-        extra=None,
-    ):
+    class _GroupingData(TypedDict):
         """
-        Upsert a Grouping generating the authority_provided_id based on its parent.
-
-        :param tool_consumer_instance_guid: Tool consumer GUID
-        :param lms_id: ID of this grouping on the LMS
-        :param lms_name: Name of the grouping on the LMS
-        :param parent: Parent of grouping
-        :param type_: Type of the grouping
-        :param extra: Any extra information to store linked to this grouping
+        Params for bulk_upsert_with_parent.
+
+        tool_consumer_instance_guid: Tool consumer GUID
+        lms_id: ID of this grouping on the LMS
+        lms_name: Name of the grouping on the LMS
+        param parent: Parent of grouping
+        type_: Type of the grouping
+        extra: Any extra information to store linked to this grouping
         """
-        authority_provided_id = self.generate_authority_provided_id(
-            tool_consumer_instance_guid, lms_id, parent, type_
-        )
 
-        grouping = (
-            self._db.query(Grouping)
-            .filter_by(
-                application_instance=self._application_instance,
-                authority_provided_id=authority_provided_id,
+        tool_consumer_instance_guid: str
+        lms_id: str
+        lms_name: str
+        parent: Grouping
+        type: Grouping.Type
+        extra: Optional[dict]
+
+    def bulk_upsert_with_parent(self, groupings_data: List[_GroupingData]):
+        """Upsert a Grouping generating the authority_provided_id based on its parent."""
+        for grouping_data in groupings_data:
+            authority_provided_id = self.generate_authority_provided_id(
+                grouping_data["tool_consumer_instance_guid"],
+                grouping_data["lms_id"],
+                grouping_data["parent"],
+                grouping_data["type"],
             )
-            .one_or_none()
-        )
 
-        if not grouping:
-            grouping = Grouping(
-                application_instance=self._application_instance,
-                authority_provided_id=authority_provided_id,
-                lms_id=lms_id,
-                parent_id=parent.id,
-                type=type_,
-            )
-            self._db.add(grouping)
+            grouping_data["authority_provided_id"] = authority_provided_id
+            grouping_data["application_instance_id"] = self._application_instance.id
+            grouping_data["parent_id"] = grouping_data["parent"].id
 
-        grouping.lms_name = lms_name
-        grouping.extra = extra
+            del grouping_data["parent"]
+            del grouping_data["tool_consumer_instance_guid"]
```

Comment:
This is not great, instead of modifying the original dict I think it would be better to crate a new dict with just the required keys.